eknrenderer/eknrenderer/eknr-renderer.c

```diff
@@ -0,0 +1,1408 @@
+  const gchar *last_section_variable = NULL;
+
+  if (value == NULL)
+    {
+      g_autofree gchar *msg = g_strdup_printf ("No such section %s", token->name);
+      (*api->error) (api, data, __LINE__, msg);
+      return 0;
+    }
+
+  if (!_renderer_section_from_ht_variant (value, api, userdata, token))
+    return 0;
+
+  return 1;
+}
+
+static void
+_renderer_set_error (mustache_api_t *api,
+                     void           *userdata,
+                     uintmax_t       lineno,
+                     const char     *msg)
+{
+  RendererMustacheData *data = userdata;
+
+  g_set_error (data->error,
+               EKNR_ERROR,
+               EKNR_ERROR_SUBSTITUTION_FAILED,
+               "Failed to perform template substitution: %s",
+               msg);
+}
+
+gchar *
+_renderer_render_mustache_document_internal (mustache_template_t  *tmpl,
+                                             GVariant             *variables,
+                                             GError              **error)
+{
+  mustache_api_t api = {
+    .read = &_renderer_read_from_closure,
+    .write = &_renderer_write_to_closure,
+    .varget = &_renderer_var_from_ht,
+    .sectget = &_renderer_sect_from_ht,
+    .error = &_renderer_set_error
+  };
+  g_autoptr(RendererMustacheData) data = renderer_mustache_data_new (g_variant_dict_new (variables),
+                                                                     error,
+                                                                     NULL);
+
+  if (!mustache_render (&api, (void *) data, tmpl))
+    return NULL;
+
+  return g_steal_pointer (&data->output.string);
+}
+
+/**
+ * eknr_renderer_render_mustache_document_from_file:
+ * @renderer: An #EknrRenderer
+ * @file: A #GFile specifying the location of the template file
+ * @variables: The variables and sections to use when rendering.
+ * @error: A #GError
+ *
+ * Use mustache_c to render a document, similar to
+ * eknr_renderer_render_mustache_document, but read the template
+ * from the file specified at @file. If that file has already been
+ * read, its contents will be read from the internal cache.
+ *
+ * Returns: The renderered document on success, %NULL on error.
+ */
+gchar *
+eknr_renderer_render_mustache_document_from_file (EknrRenderer *renderer,
+                                                  GFile        *file,
+                                                  GVariant     *variables,
+                                                  GError      **error)
+{
+  g_autofree gchar *path = g_file_get_uri (file);
+  EknrRendererPrivate *priv = eknr_renderer_get_instance_private (renderer);
+  mustache_template_t *tmpl = g_hash_table_lookup (priv->cache, path);
+  g_autofree gchar *contents = NULL;
+  mustache_api_t api = {
+    .read = &_renderer_read_from_closure,
+    .write = &_renderer_write_to_closure,
+    .varget = &_renderer_var_from_ht,
+    .sectget = &_renderer_sect_from_ht,
+    .error = &_renderer_set_error
+  };
```

Comment:
I note that this is the same vtable as above, so it could even be a compilation-unit-global...